### PR TITLE
create DepartmentService

### PR DESCRIPTION
**Pull request by Cheephot**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/906

### CODE REVIEW

File: src/DepartmentService.Broker/GetDepartmentsConsumer.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  {
+    private readonly IDepartmentRepository _repository;
+    private readonly IConnectionMultiplexer _cache;
+    private readonly IOptions<RedisConfig> _redisConfig;
+
+    private async Task<List<DepartmentData>> GetDepartment(IGetDepartmentsRequest request)
+    {
+      List<DbDepartment> dbDepartments = new();
+
+      List<Guid> departmentIds = request.DepartmentsIds;
```
Comment (MariyaSin) [inline src/DepartmentService.Broker/GetDepartmentsConsumer.cs:32]: departmentS

File: src/DepartmentService.Broker/GetDepartmentsConsumer.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    public async Task Consume(ConsumeContext<IGetDepartmentsRequest> context)
+    {
+      List<DepartmentData> departments = await GetDepartment(context.Message);
+
+      object departmentId = OperationResultWrapper.CreateResponse((_) => IGetDepartmentsResponse.CreateObj(departments), context);
+
+      await context.RespondAsync<IOperationResult<IGetDepartmentsResponse>>(departmentId);
+
+      await _cache.GetDatabase(Cache.Departments).StringSetAsync(
```
Comment (BelehovEgor) [inline src/DepartmentService.Broker/GetDepartmentsConsumer.cs:65]: you can use redis helper

File: src/DepartmentService.Business/AddDepartmentUsersCommand.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+      if (!(await _accessValidator.HasRightsAsync(Rights.AddEditRemoveDepartments) &&
+        await _repository.IsDepartmentDirectorAsync(request.DepartmentId, senderUserId)))
+      {
+        _httpContextAccessor.HttpContext.Response.StatusCode = (int)HttpStatusCode.Forbidden;
+
+        return new OperationResultResponse<bool>
+        {
+          Status = OperationResultStatusType.Failed,
+          Errors = new() { "Not enough rights." }
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/DepartmentService.Business/AddDepartmentUsersCommand.cs:54]: can use responseCreater

File: src/DepartmentService.Business/GetDepartmentCommand.cs [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+    private readonly IConnectionMultiplexer _cache;
+
+    private async Task<List<UserData>> GetUsersDataAsync(List<Guid> usersIds, List<string> errors)
+    {
+      if (usersIds == null || !usersIds.Any())
+      {
+        return new();
+      }
+
+      RedisValue valueFromCache = await _cache.GetDatabase(Cache.Users).StringGetAsync(usersIds.GetRedisCacheHashCode());
```
Comment (BelehovEgor) [inline src/DepartmentService.Business/GetDepartmentCommand.cs:53]: you can user redis helper

File: src/DepartmentService.Data.Provider.MsSql.Ef/Migrations/20211012014055_InitialCreate.cs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+          CreatedBy = table.Column<Guid>(nullable: false),
+          CreatedAtUtc = table.Column<DateTime>(nullable: false),
+          ModifiedBy = table.Column<Guid>(nullable: true),
+          ModifiedAtUtc = table.Column<DateTime>(nullable: true),
+          LeftAtUtc = table.Column<DateTime>(nullable: true)
+        },
+        constraints: table =>
+        {
+          table.PrimaryKey("PK_DepartmentUser", x => x.Id);
+          table.ForeignKey(
```
Comment (BelehovEgor) [inline src/DepartmentService.Data.Provider.MsSql.Ef/Migrations/20211012014055_InitialCreate.cs:58]: remove

Comment (BelehovEgor) [reply src/DepartmentService.Data.Provider.MsSql.Ef/Migrations/20211012014055_InitialCreate.cs:58]: we don't use FK here

File: src/DepartmentService.Data/Interfaces/IDepartmentUserRepository.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+    Task<bool> AddAsync(List<DbDepartmentUser> departmentUsers);
+
+    Task<DbDepartmentUser> GetAsync(Guid userId, bool includeDepartment);
+
+    List<Guid> Get(IGetDepartmentUsersRequest request, out int totalCount);
+
+    Task<List<DbDepartmentUser>> GetAsync(List<Guid> userIds);
+
+    void RemoveAsync(Guid userId, Guid removedBy);
```
Comment (BelehovEgor) [inline src/DepartmentService.Data/Interfaces/IDepartmentUserRepository.cs:23]: user Task

File: src/DepartmentService.Mappers/Responses/DepartmentResponseMapper.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+    public DepartmentResponse Map(
+      DbDepartment dbDepartment,
+      List<UserData> usersData,
+      List<PositionData> positionData,
+      List<ImageData> userImages,
+      List<ProjectData> projectsInfo,
+      GetDepartmentFilter filter)
+    {
+
```
Comment (BelehovEgor) [inline src/DepartmentService.Mappers/Responses/DepartmentResponseMapper.cs:53]: remove line

File: src/DepartmentService.Models.Db/DbDepartmentProject.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+      builder
+        .ToTable(DbDepartmentProject.TableName);
+
+      builder
+        .HasKey(u => u.Id);
+
+      builder
+        .HasOne(u => u.Department)
+        .WithMany(d => d.Projects)
+        .HasForeignKey(u => u.DepartmentId);
```
Comment (BelehovEgor) [inline src/DepartmentService.Models.Db/DbDepartmentProject.cs:34]: you can don't write it

File: src/DepartmentService.Models.Db/DbDepartmentUser.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+      builder
+        .ToTable(DbDepartmentUser.TableName);
+
+      builder
+        .HasKey(u => u.Id);
+
+      builder
+        .HasOne(u => u.Department)
+        .WithMany(d => d.Users)
+        .HasForeignKey(u => u.DepartmentId);
```
Comment (BelehovEgor) [inline src/DepartmentService.Models.Db/DbDepartmentUser.cs:37]: same

File: src/DepartmentService.Business/AddDepartmentUsersCommand.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Cheephot) [reply src/DepartmentService.Business/AddDepartmentUsersCommand.cs:54]: done

File: src/DepartmentService.Data/Interfaces/IDepartmentUserRepository.cs (hunk 6 above)
Comment (Cheephot) [reply src/DepartmentService.Data/Interfaces/IDepartmentUserRepository.cs:23]: why?

File: src/DepartmentService.Models.Db/DbDepartmentProject.cs (hunk 8 above)
Comment (Cheephot) [reply src/DepartmentService.Models.Db/DbDepartmentProject.cs:34]: done

File: .github/workflows/main.yml [hunk 10]
```diff
@@ -0,0 +1,36 @@
+name: Build & Test DepartmentService
+
+# Controls when the action will run. Triggers the workflow on push or pull request
+# events but only for the develop branch
+on:
+  push:
+    branches: [ '*' ]
+  pull_request:
+    branches: [ develop, master, 'release/*' ]
```
[the remark this answers is not in the document]
Comment (Cheephot) [reply .github/workflows/main.yml:9]: done

File: src/DepartmentService.Broker/GetDepartmentsConsumer.cs (hunk 2 above)
Comment (Cheephot) [reply src/DepartmentService.Broker/GetDepartmentsConsumer.cs]: done

File: src/DepartmentService.Data/Interfaces/IDepartmentUserRepository.cs (hunk 6 above)
Comment (BelehovEgor) [reply src/DepartmentService.Data/Interfaces/IDepartmentUserRepository.cs:23]: we can use void with async methods, but it works differently in one moment. 
https://docs.microsoft.com/ru-ru/dotnet/csharp/programming-guide/concepts/async/async-return-types

File: src/DepartmentService.Data/DepartmentUserRepository.cs [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+    public async Task<List<DbDepartmentUser>> GetAsync(List<Guid> userIds)
+    {
+      return await _provider.DepartmentUsers
+        .Include(du => du.Department)
+        .Where(u => u.IsActive && userIds.Contains(u.UserId))
+        .ToListAsync();
+    }
+
+    public async void RemoveAsync(Guid userId, Guid removedBy)
```
Comment (BelehovEgor) [inline src/DepartmentService.Data/DepartmentUserRepository.cs:146]: use Task

File: src/DepartmentService.Data/DepartmentUserRepository.cs [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+        dbDepartmentUser.IsActive = false;
+        dbDepartmentUser.ModifiedAtUtc = DateTime.UtcNow;
+        dbDepartmentUser.ModifiedBy = removedBy;
+        dbDepartmentUser.LeftAtUts = DateTime.UtcNow;
+
+        await _provider.SaveAsync();
+      }
+    }
+
+    public async void RemoveAsync(List<Guid> usersIds, Guid removedBy)
```
Comment (BelehovEgor) [inline src/DepartmentService.Data/DepartmentUserRepository.cs:162]: use Task

File: src/DepartmentService.Data/Interfaces/IDepartmentRepository.cs [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+
+    Task<List<DbDepartment>> GetAsync(List<Guid> departmentsIds, bool includeUsers = false);
+
+    Task<DbDepartment> GetAsync(GetDepartmentFilter filter);
+
+    Task<(List<DbDepartment>, int totalCount)> FindAsync(FindDepartmentFilter filter);
+
+    Task<bool> EditAsync(DbDepartment department, JsonPatchDocument<DbDepartment> request);
+
+    List<DbDepartment> Search(string text);
```
Comment (BelehovEgor) [inline src/DepartmentService.Data/Interfaces/IDepartmentRepository.cs:24]: make async

File: src/DepartmentService.Mappers/Db/DbDepartmentNewsMapper.cs [hunk 14]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using LT.DigitalOffice.DepartmentService.Mappers.Db.Interfaces;
+using LT.DigitalOffice.DepartmentService.Models.Db;
+
+namespace LT.DigitalOffice.DepartmentService.Mappers.Db
+{
+  public class DbDepartmentNewsMapper : IDbDepartmentNewsMapper
+  {
+    public DbDepartmentNews Map(Guid newsId, Guid departmentId, Guid modifiedBy)
```
Comment (BelehovEgor) [inline src/DepartmentService.Mappers/Db/DbDepartmentNewsMapper.cs:9]: maybe here better createdBy instead modifiedBy

File: src/DepartmentService.Mappers/Db/DbDepartmentProjectMapper.cs [hunk 15]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using LT.DigitalOffice.DepartmentService.Mappers.Db.Interfaces;
+using LT.DigitalOffice.DepartmentService.Models.Db;
+
+namespace LT.DigitalOffice.DepartmentService.Mappers
+{
+  public class DbDepartmentProjectMapper : IDbDepartmentProjectMapper
+  {
+    public DbDepartmentProject Map(Guid projectId, Guid departmentId, Guid modifiedBy)
```
Comment (BelehovEgor) [inline src/DepartmentService.Mappers/Db/DbDepartmentProjectMapper.cs:9]: same

File: src/DepartmentService.Mappers/Db/DbDepartmentUserMapper.cs [hunk 16]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using LT.DigitalOffice.DepartmentService.Mappers.Db.Interfaces;
+using LT.DigitalOffice.DepartmentService.Models.Db;
+
+namespace LT.DigitalOffice.DepartmentService.Mappers.Db
+{
+  public class DbDepartmentUserMapper : IDbDepartmentUserMapper
+  {
+    public DbDepartmentUser Map(Guid userId, Guid departmentId, Guid modifiedBy)
```
Comment (BelehovEgor) [inline src/DepartmentService.Mappers/Db/DbDepartmentUserMapper.cs:9]: same

File: src/DepartmentService.Mappers/Models/DepartmentUserInfoMapper.cs [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+
+      return new UserInfo
+      {
+        Id = userData.Id,
+        FirstName = userData.FirstName,
+        LastName = userData.LastName,
+        MiddleName = userData.MiddleName,
+        Rate = userData.Rate,
+        IsActive = userData.IsActive,
+        Image = image != null ? _imageMapper.Map(image) : null,
```
Comment (BelehovEgor) [inline src/DepartmentService.Mappers/Models/DepartmentUserInfoMapper.cs:36]: now you can write _imageMapper.Map(image) ofc if mapper doesn't throw exc

File: src/DepartmentService.Mappers/Models/DepartmentUserInfoMapper.cs [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+        Position = positionData != null ? _positionInfoMapper.Map(positionData) : null
```
Comment (BelehovEgor) [inline src/DepartmentService.Mappers/Models/DepartmentUserInfoMapper.cs:37]: same

File: src/DepartmentService.Models.Db/DbDepartment.cs [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+
+  public class DbDepartmentConfiguration : IEntityTypeConfiguration<DbDepartment>
+  {
+    public void Configure(EntityTypeBuilder<DbDepartment> builder)
+    {
+      builder
+        .ToTable(DbDepartment.TableName);
+
+      builder
+        .HasKey(d => d.Id);
```
Comment (BelehovEgor) [inline src/DepartmentService.Models.Db/DbDepartment.cs:44]: add that name is required property

File: src/DepartmentService.Models.Db/DbDepartmentNews.cs [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+
+    public Guid Id { get; set; }
+    public Guid NewsId { get; set; }
+    public Guid DepartmentId { get; set; }
+    public bool IsActive { get; set; }
+    public Guid CreatedBy { get; set; }
+    public DateTime CreatedAtUtc { get; set; }
+    public Guid? ModifiedBy { get; set; }
+    public DateTime? ModifiedAtUtc { get; set; }
+    public DbDepartment Department { get; set; }
```
Comment (BelehovEgor) [inline src/DepartmentService.Models.Db/DbDepartmentNews.cs:19]: add line before

File: src/DepartmentService.Models.Db/DbDepartmentProject.cs [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+    public const string TableName = "DepartmentsProjects";
+
+    public Guid Id { get; set; }
+    public Guid ProjectId { get; set; }
+    public Guid DepartmentId { get; set; }
+    public Guid CreatedBy { get; set; }
+    public DateTime CreatedAtUtc { get; set; }
+    public Guid? ModifiedBy { get; set; }
+    public DateTime? ModifiedAtUtc { get; set; }
+    public DbDepartment Department { get; set; }
```
Comment (BelehovEgor) [inline src/DepartmentService.Models.Db/DbDepartmentProject.cs:18]: add line before

File: src/DepartmentService.Models.Db/DbDepartmentUser.cs [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+    public Guid Id { get; set; }
+    public Guid UserId { get; set; }
+    public Guid DepartmentId { get; set; }
+    public int Role { get; set; }
+    public bool IsActive { get; set; }
+    public Guid CreatedBy { get; set; }
+    public DateTime CreatedAtUtc { get; set; }
+    public Guid? ModifiedBy { get; set; }
+    public DateTime? ModifiedAtUtc { get; set; }
+    public DateTime? LeftAtUts { get; set; }
```
Comment (BelehovEgor) [inline src/DepartmentService.Models.Db/DbDepartmentUser.cs:20]: Utc

File: src/DepartmentService.Models.Db/DbDepartmentUser.cs [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+    public DbDepartment Department { get; set; }
```
Comment (BelehovEgor) [inline src/DepartmentService.Models.Db/DbDepartmentUser.cs:21]: add line

File: src/DepartmentService.Models.Dto/Configurations/RabbitMqConfig.cs [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace LT.DigitalOffice.DepartmentService.Models.Dto.Configuration
+{
+  public class RabbitMqConfig : BaseRabbitMqConfig
+  {
+    public string GetDepartmentsEndpoint { get; set; }
+    public string GetDepartmentUsersEndpoint { get; set; }
+    public string SearchDepartmentEndpoint { get; set; }
+
+    [AutoInjectRequest(typeof(IGetUsersDataRequest))]
```
Comment (BelehovEgor) [inline src/DepartmentService.Models.Dto/Configurations/RabbitMqConfig.cs:18]: It is not necessary but, maybe better to write requests from one service closely

File: src/DepartmentService/appsettings.json [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+    ]
+  },
+  "ServiceInfo": {
+    "Id": "AB21C02C-C960-4C3E-8D63-1B570CFD531E",
+    "Name": "DepartmentService"
+  },
+  "ConnectionStrings": {
+    "Redis": "localhost:6379"
+  },
+  "EmailEngineConfig": {
```
Comment (BelehovEgor) [inline src/DepartmentService/appsettings.json:32]: why is it here?

File: src/DepartmentService/appsettings.json [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+    // CompanyService
+    "GetPositionsEndpoint": "CompanyService_GetPositions",
+
+    // DepartmentService
+    "GetDepartmentsEndpoint": "DepartmentService_GetDepartments",
+    "GetDepartmentUsersEndpoint": "DepartmentService_GetDepartmentUsers",
+    "SearchDepartmentEndpoint": "DepartmentService_SearchDepartment"
+  },
+  "CheckTokenMiddleware": {
+    "SkippedEndpoints": [
```
Comment (BelehovEgor) [inline src/DepartmentService/appsettings.json:69]: add api info

File: src/DepartmentService.Data/Interfaces/IDepartmentUserRepository.cs (hunk 6 above)
Comment (Cheephot) [reply src/DepartmentService.Data/Interfaces/IDepartmentUserRepository.cs:23]: thx, done

File: src/DepartmentService.Business/CreateDepartmentCommand.cs [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+        return new OperationResultResponse<Guid>
+        {
+          Status = OperationResultStatusType.Failed,
+          Errors = errors
+        };
+      }
+
+      OperationResultResponse<Guid> response = new();
+
+      if (await _repository.DoesNameExistAsync(request.Name))
```
Comment (MariyaSin) [inline src/DepartmentService.Business/CreateDepartmentCommand.cs:71]: remove to validator

File: src/DepartmentService.Business/CreateDepartmentUsersCommand.cs [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+        return new OperationResultResponse<bool>
+        {
+          Status = OperationResultStatusType.Failed,
+          Errors = validationResult.Errors.Select(vf => vf.ErrorMessage).ToList()
+        };
+      }
+
+      OperationResultResponse<bool> response = new();
+
+      _repository.RemoveAsync(request.UsersIds, senderUserId);
```
Comment (MariyaSin) [inline src/DepartmentService.Business/CreateDepartmentUsersCommand.cs:71]: await

Comment (MariyaSin) [reply src/DepartmentService.Business/CreateDepartmentUsersCommand.cs:71]: you can get senderUserId in repo from HTTP conext

File: src/DepartmentService.Business/EditDepartmentCommand.cs [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+        _httpContextAccessor.HttpContext.Response.StatusCode = (int)HttpStatusCode.BadRequest;
+
+        return new OperationResultResponse<bool>
+        {
+          Status = OperationResultStatusType.Failed,
+          Errors = errors
+        };
+      }
+
+      DbDepartment dbDepartment = await _repository.GetAsync(new GetDepartmentFilter { DepartmentId = departmentId });
```
Comment (MariyaSin) [inline src/DepartmentService.Business/EditDepartmentCommand.cs:73]: for what do you get dbDepartment in command?

File: src/DepartmentService.Business/EditDepartmentCommand.cs [hunk 30]
```diff
@@ -0,0 +1,115 @@
...
+      }
+
+      OperationResultResponse<bool> response = new();
+
+      foreach (Operation<EditDepartmentRequest> item in request.Operations)
+      {
+        if (item.path.EndsWith(nameof(EditDepartmentRequest.Name), StringComparison.OrdinalIgnoreCase) &&
+            await _repository.DoesNameExistAsync(item.value.ToString()))
+        {
+          _httpContextAccessor.HttpContext.Response.StatusCode = (int)HttpStatusCode.Conflict;
```
Comment (MariyaSin) [inline src/DepartmentService.Business/EditDepartmentCommand.cs:93]: куызщтыу скуфещк

File: src/DepartmentService.Business/FindDepartmentCommand.cs [hunk 31]
```diff
@@ -0,0 +1,246 @@
...
+
+    private async Task<List<PositionData>> GetPositionAsync(List<Guid> positionIds, List<string> errors)
+    {
+      if (positionIds == null || !positionIds.Any())
+      {
+        return new();
+      }
+
+      string logMessage = "Can not get position: {ids}.";
+      string errorMessage = "Can not get positions. Please try again later.";
```
Comment (MariyaSin) [inline src/DepartmentService.Business/FindDepartmentCommand.cs:144]: you use the meesage only in 164 string. No need create it here

File: src/DepartmentService.Business/FindDepartmentCommand.cs [hunk 32]
```diff
@@ -0,0 +1,246 @@
...
+      _rcGetImages = rcGetImages;
+      _rcGetPositions = rcGetPosition;
+      _logger = logger;
+      _httpContextAccessor = httpContextAccessor;
+      _cache = cache;
+    }
+
+    public async Task<FindResultResponse<DepartmentInfo>> ExecuteAsync(FindDepartmentFilter filter)
+    {
+
```
Comment (MariyaSin) [inline src/DepartmentService.Business/FindDepartmentCommand.cs:195]: remove empty line

File: src/DepartmentService.Business/FindDepartmentCommand.cs [hunk 33]
```diff
@@ -0,0 +1,246 @@
...
+      FindResultResponse<DepartmentInfo> response = new(body: new());
+
+      (List<DbDepartment> dbDepartments, int totalCount) = await _repository.FindAsync(filter);
+
+      if (!_baseFindValidator.ValidateCustom(filter, out List<string> errors))
+      {
+        _httpContextAccessor.HttpContext.Response.StatusCode = (int)HttpStatusCode.BadRequest;
```
Comment (MariyaSin) [inline src/DepartmentService.Business/FindDepartmentCommand.cs:202]: response ceator

File: src/DepartmentService.Business/GetDepartmentCommand.cs [hunk 34]
```diff
@@ -0,0 +1,276 @@
...
+      {
+        (List<ProjectData> projects, int _) = JsonConvert.DeserializeObject<(List<ProjectData>, int)>(projectsFromCache);
+
+        return projects;
+      }
+
+      return await GetProjectsDatasThroughBroker(departmentId, errors);
+    }
+
+    private async Task<List<ProjectData>> GetProjectsDatasThroughBroker(Guid departmentId, List<string> errors)
```
Comment (MariyaSin) [inline src/DepartmentService.Business/GetDepartmentCommand.cs:117]: GetProjectsDatasThroughBrokerAsync

File: src/DepartmentService.Business/GetDepartmentCommand.cs [hunk 35]
```diff
@@ -0,0 +1,276 @@
...
+    {
+      string message = "Can not get projects data. Please try again later.";
```
Comment (MariyaSin) [inline src/DepartmentService.Business/GetDepartmentCommand.cs:119]: you use the message only in 140 line. do not need create it here

File: src/DepartmentService.Data/DepartmentRepository.cs [hunk 36]
```diff
@@ -0,0 +1,118 @@
...
+      {
+        dbDepartments = dbDepartments.Include(d => d.Users.Where(u => u.IsActive));
+      }
+
+      return await dbDepartments.FirstOrDefaultAsync();
+    }
+
+    public async Task<(List<DbDepartment>, int totalCount)> FindAsync(FindDepartmentFilter filter)
+    {
+      return (
```
Comment (MariyaSin) [inline src/DepartmentService.Data/DepartmentRepository.cs:54]: check if filter is null

File: src/DepartmentService.Data/DepartmentUserRepository.cs [hunk 37]
```diff
@@ -0,0 +1,181 @@
...
+        return false;
+      }
+
+      _provider.DepartmentUsers.Add(departmentUser);
+      await _provider.SaveAsync();
+
+      return true;
+    }
+
+    public async Task<bool> AddAsync(List<DbDepartmentUser> departmentUsers)
```
Comment (MariyaSin) [inline src/DepartmentService.Data/DepartmentUserRepository.cs:37]: CreateAsync, departmentsUsers

File: src/DepartmentService.Data/DepartmentUserRepository.cs [hunk 38]
```diff
@@ -0,0 +1,181 @@
...
+
+      if (request.TakeCount.HasValue)
+      {
+        dbDepartmentUser = dbDepartmentUser.Take(request.TakeCount.Value);
+      }
+
+      return (await dbDepartmentUser.Select(x => x.UserId).ToListAsync(), totalCount);
+    }
+
+    public async Task<List<DbDepartmentUser>> GetAsync(List<Guid> userIds)
```
Comment (MariyaSin) [inline src/DepartmentService.Data/DepartmentUserRepository.cs:138]: usersIds
Check if usersIds == null

File: src/DepartmentService.Mappers/Db/DbDepartmentMapper.cs [hunk 39]
```diff
@@ -0,0 +1,64 @@
...
+
+      DbDepartment dbDepartment = new DbDepartment
+      {
+        Id = departmentId,
+        Name = request.Name,
+        Description = request.Description,
+        IsActive = true,
+        CreatedBy = authorId,
+        CreatedAtUtc = DateTime.UtcNow,
+        Users = request.Users?.Select(du => GetDbDepartmentUser(departmentId, du, authorId, DepartmentUserRole.Employee)).ToList() ?? new()
```
Comment (MariyaSin) [inline src/DepartmentService.Mappers/Db/DbDepartmentMapper.cs:39]: can you use DepartmentUserMapper here?

File: src/DepartmentService/appsettings.json [hunk 40]
```diff
@@ -0,0 +1,73 @@
...
+    "CheckUserRightsEndpoint": "CRService_CheckUserRights",
+
+    // AuthService
+    "ValidateTokenEndpoint": "AuthService_ValidateToken",
+
+    // ProjectService
+    "GetProjectsEndpoint": "ProjectService_GetProjects",
+
+    // MessageService
+    "UpdateSmtpCredentialsEndpoint": "MessageService_UpdateSmtpCredentials",
```
Comment (MariyaSin) [inline src/DepartmentService/appsettings.json:55]: do you use the endpoint?

File: src/DepartmentService/appsettings.json [hunk 41]
```diff
@@ -0,0 +1,73 @@
...
+    "CheckUsersExistenceEndpoint": "UserService_CheckUsersExistence",
+
+    // CRService
+    "CheckUserRightsEndpoint": "CRService_CheckUserRights",
+
+    // AuthService
+    "ValidateTokenEndpoint": "AuthService_ValidateToken",
+
+    // ProjectService
+    "GetProjectsEndpoint": "ProjectService_GetProjects",
```
Comment (MariyaSin) [inline src/DepartmentService/appsettings.json:52]: do you use it?

File: src/DepartmentService/Startup.cs [hunk 42]
```diff
@@ -0,0 +1,259 @@
...
+      {
+        ep.ConfigureConsumer<SearchDepartmentsConsumer>(context);
+      });
+      cfg.ReceiveEndpoint(_rabbitMqConfig.GetDepartmentUsersEndpoint, ep =>
+      {
+        ep.ConfigureConsumer<GetDepartmentUsersConsumer>(context);
+      });
+    }
+
+    private string HidePassword(string line)
```
Comment (MariyaSin) [inline src/DepartmentService/Startup.cs:223]: use kernel

File: src/DepartmentService.Validation/EditDepartmentRequestValidator.cs [hunk 43]
```diff
@@ -0,0 +1,96 @@
...
+      #endregion
+
+      #region Name
+
+      AddFailureForPropertyIf(
+        nameof(EditDepartmentRequest.Name),
+        x => x == OperationType.Replace,
+        new()
+        {
+          { x => !string.IsNullOrEmpty(x.value?.ToString().Trim()), "Name must not be empty." },
```
Comment (MariyaSin) [inline src/DepartmentService.Validation/EditDepartmentRequestValidator.cs:45]: null ref with trim?

File: src/DepartmentService.Validation/EditDepartmentRequestValidator.cs [hunk 44]
```diff
@@ -0,0 +1,96 @@
...
+      #endregion
+
+      #region Description
+
+      AddFailureForPropertyIf(
+        nameof(EditDepartmentRequest.Description),
+        x => x == OperationType.Replace || x == OperationType.Add,
+        new()
+        {
+          { x => !string.IsNullOrEmpty(x.value?.ToString().Trim()), "Description must not be empty." },
```
Comment (MariyaSin) [inline src/DepartmentService.Validation/EditDepartmentRequestValidator.cs:59]: is it required property?

File: src/DepartmentService.Validation/EditDepartmentRequestValidator.cs [hunk 45]
```diff
@@ -0,0 +1,96 @@
...
+      #endregion
+
+      #region DirectorId
+
+      AddFailureForPropertyIf(
+        nameof(EditDepartmentRequest.DirectorId),
+        x => x == OperationType.Replace,
+        new()
+        {
+          { x => Guid.TryParse(x.value?.ToString(), out Guid _), "Incorrect format of DirectorId." },
```
Comment (MariyaSin) [inline src/DepartmentService.Validation/EditDepartmentRequestValidator.cs:72]: check if default value

File: src/DepartmentService.Models.Dto/Requests/Filters/GetDepartmentFilter.cs [hunk 46]
```diff
@@ -0,0 +1,20 @@
...
+  public record GetDepartmentFilter
+  {
+    [FromQuery(Name = "departmentid")]
+    public Guid DepartmentId { get; set; }
+
+    [FromQuery(Name = "includeusers")]
+    public bool? IncludeUsers { get; set; }
+
+    [FromQuery(Name = "includeprojects")]
+    public bool? IncludeProjects { get; set; }
```
Comment (MariyaSin) [inline src/DepartmentService.Models.Dto/Requests/Filters/GetDepartmentFilter.cs:15]: public bool IncludeProjects { get; set; } = false

File: src/DepartmentService.Models.Dto/Requests/Filters/GetDepartmentFilter.cs [hunk 47]
```diff
@@ -0,0 +1,20 @@
...
+
+    public bool IsIncludeUsers => IncludeUsers.HasValue && IncludeUsers.Value;
```
Comment (MariyaSin) [inline src/DepartmentService.Models.Dto/Requests/Filters/GetDepartmentFilter.cs:17]: remove

File: src/DepartmentService.Models.Dto/Requests/Filters/GetDepartmentFilter.cs [hunk 48]
```diff
@@ -0,0 +1,20 @@
...
+    public bool IsIncludeProjects => IncludeProjects.HasValue && IncludeProjects.Value;
```
Comment (MariyaSin) [inline src/DepartmentService.Models.Dto/Requests/Filters/GetDepartmentFilter.cs:18]: remove

File: src/DepartmentService.Mappers/Db/DbDepartmentProjectMapper.cs [hunk 49]
```diff
@@ -0,0 +1,30 @@
...
+namespace LT.DigitalOffice.DepartmentService.Mappers.Db
+{
+  public class DbDepartmentProjectMapper : IDbDepartmentProjectMapper
+  {
+    private readonly IHttpContextAccessor _httpContextAccessor;
+
+    public DbDepartmentProjectMapper(IHttpContextAccessor httpContextAccessor)
+    {
+      _httpContextAccessor = httpContextAccessor;
+    }
```
Comment (BelehovEgor) [inline src/DepartmentService.Mappers/Db/DbDepartmentProjectMapper.cs:16]: add line

File: src/DepartmentService.Mappers/Responses/Interfaces/IDepartmentResponseMapper.cs [hunk 50]
```diff
@@ -0,0 +1,17 @@
...
+using LT.DigitalOffice.DepartmentService.Models.Dto.Responses;
+using LT.DigitalOffice.Kernel.Attributes;
+
+namespace LT.DigitalOffice.DepartmentService.Mappers.Responses.Interfaces
+{
+  [AutoInject]
+  public interface IDepartmentResponseMapper
+  {
+    DepartmentResponse Map(
+          DbDepartment dbDepartment,
```
Comment (BelehovEgor) [inline src/DepartmentService.Mappers/Responses/Interfaces/IDepartmentResponseMapper.cs:13]: fix tabs

File: src/DepartmentService.Validation/EditDepartmentRequestValidator.cs [hunk 51]
```diff
@@ -0,0 +1,95 @@
...
+          { x => x.value.ToString().Trim().Length > 2, "Name is too short." },
+        }, CascadeMode.Stop);
+
+      #endregion
+
+      #region Description
+
+      AddFailureForPropertyIf(
+        nameof(EditDepartmentRequest.Description),
+        x => x == OperationType.Replace || x == OperationType.Add,
```
Comment (BelehovEgor) [inline src/DepartmentService.Validation/EditDepartmentRequestValidator.cs:56]: remove add

**Review by BelehovEgor — APPROVED** [review]

**Pull request merged**